From afba25094274772e2b3454ce9897a47e4d271818 Mon Sep 17 00:00:00 2001
From: Subbaraya Sundeep <sbhatta@marvell.com>
Date: Fri, 15 Mar 2019 20:30:13 +0530
Subject: [PATCH 081/386] octeontx2-pf: Use user-def parameter of ethtool
 ntuple command

user-def parameter of ethtool --config-ntuple command is for
adding any vendor specific options. This patch uses user-def
for specifying action other than Drop and Direct to queue.
When user-def is 0x1 then action of default unicast entry
is used which inturn is a RSS action.
Example:
ethtool -U eth2 flow-type udp4 src-port 23 action 0 user-def 0x1

Change-Id: I106b95ccd69e9a7c8e113da56fa58fcbde96ed53
Signed-off-by: Subbaraya Sundeep <sbhatta@marvell.com>
Reviewed-on: https://sj1git1.cavium.com/6349
Tested-by: sa_ip-sw-jenkins
Reviewed-by: Sunil Kovvuri Goutham <Sunil.Goutham@cavium.com>
[RH: Original patch taken from marvell 88F3720 board support SDK 10.0-PR2003]
Signed-off-by: Ruiqiang Hao <Ruiqiang.Hao@windriver.com>
---
 .../ethernet/marvell/octeontx2/nic/otx2_ethtool.c    | 12 +++++++++++-
 1 file changed, 11 insertions(+), 1 deletion(-)

diff --git a/drivers/net/ethernet/marvell/octeontx2/nic/otx2_ethtool.c b/drivers/net/ethernet/marvell/octeontx2/nic/otx2_ethtool.c
index 5d2ca27cc1d7..6050773cb67e 100644
--- a/drivers/net/ethernet/marvell/octeontx2/nic/otx2_ethtool.c
+++ b/drivers/net/ethernet/marvell/octeontx2/nic/otx2_ethtool.c
@@ -23,6 +23,8 @@
 #define DRV_VF_NAME	"octeontx2-nicvf"
 #define DRV_VF_VERSION	"1.0"
 
+#define OTX2_DEFAULT_ACTION	0x1
+
 struct otx2_stat {
 	char name[ETH_GSTRING_LEN];
 	unsigned int index;
@@ -689,6 +691,10 @@ static int otx2_prepare_flow_request(struct ethtool_rx_flow_spec *fsp,
 			       sizeof(pmask->vlan_tci));
 			req->features |= BIT_ULL(NPC_OUTER_VID);
 		}
+		/* Not Drop/Direct to queue but use action in default entry */
+		if (fsp->m_ext.data[1] &&
+		    fsp->h_ext.data[1] == cpu_to_be32(OTX2_DEFAULT_ACTION))
+			req->op = NIX_RX_ACTION_DEFAULT;
 	}
 	if (fsp->flow_type & FLOW_MAC_EXT &&
 	    !is_zero_ether_addr(fsp->m_ext.h_dest)) {
@@ -731,7 +737,11 @@ static int otx2_add_flow_msg(struct otx2_nic *pfvf, struct otx2_flow *flow)
 	if (ring_cookie == RX_CLS_FLOW_DISC) {
 		req->op = NIX_RX_ACTIONOP_DROP;
 	} else {
-		req->op = NIX_RX_ACTIONOP_UCAST;
+		/* change to unicast only if action of default entry is not
+		 * requested by user
+		 */
+		if (req->op != NIX_RX_ACTION_DEFAULT)
+			req->op = NIX_RX_ACTIONOP_UCAST;
 		req->index = ethtool_get_flow_spec_ring(ring_cookie);
 		vf = ethtool_get_flow_spec_ring_vf(ring_cookie);
 		if (vf > pci_num_vf(pfvf->pdev)) {
-- 
2.17.1

